### ISSUE TITLE
Bug: Incorrect Hydration Mismatch Detection during Suspense - "Hydration failed because the initial UI does not match what was rendered on the server."

### ISSUE THREAD

**Issue by xiel** [facebook/react#24384]
React version: 18.0.0

## Steps To Reproduce

1. Add a Suspense Boundary
2. Add a **component that will suspend** to load some data (faked).
3. Render at least one **sibling component** _after_ the suspending component.
3. Render server-side using renderToPipeableStream()
4. Render client-side using hydrateRoot()

__Reproductions in CodeSandbox:__

- [Reproduction 1](https://yt3148.sse.codesandbox.io) with Next.js [Edit](https://codesandbox.io/s/github/xiel/SuspenseHydrationErrorNext)
- [Reproduction 2](https://codesandbox.io/s/brave-euclid-lx3et7?file=/server/render.js) using renderToPipeableStream (based on [example demo](https://codesandbox.io/s/kind-sammet-j56ro?file=/server/render.js:1054-1614))

````jsx
<>
  <h4>This headline hydrates fine. </h4>
  <SomethingA />{/* Will suspend on client and server */}
  <h3>💥 This element after the suspending Component triggers an error (only in development).</h3>
</>
````

## The current behavior

- The sibling component following the suspending component is incorrectly seen as a **hydration mismatch**.
- Console will show errors (and Next.js will show big error overlay):

````
Warning: Expected server HTML to contain a matching <h3> in <div>.
    at h3
    at IndexPage
    at Suspense
    at App (webpack-internal:///./pages/_app.js:42:27)
    at ErrorBoundary (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/client.js:8:20638)
    at ReactDevOverlay (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/client.js:8:23179)
    at Container (webpack-internal:///./node_modules/next/dist/client/index.js:323:9)
    at AppContainer (webpack-internal:///./node_modules/next/dist/client/index.js:825:26)
    at Root (webpack-internal:///./node_modules/next/dist/client/index.js:949:27)
````

````
Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
    at throwOnHydrationMismatch (react-dom.development.js?ac89:14344:1)
    at tryToClaimNextHydratableInstance (react-dom.development.js?ac89:14372:1)
    at updateHostComponent$1 (react-dom.development.js?ac89:20636:1)
    at beginWork (react-dom.development.js?ac89:22373:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js?ac89:4157:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js?ac89:4206:1)
    at invokeGuardedCallback (react-dom.development.js?ac89:4270:1)
````

- Depending on the location of the suspending component and their siblings, there are errors or no errors at all. When the suspending component is the last component, there are no errors logged.
- Error can be suppressed by wrapping the suspending component in a Suspense boundary directly

## The expected behavior

- I expect to be able to use the same fetching mechanism using suspense on server and client in [React v18](https://reactjs.org/blog/2022/03/29/react-v18.html#suspense-in-data-frameworks).
_OR_
- If this is unsupported usage, the errors should be more clear and consistent.

**Comment by gaearon** [comment]
Small repro: https://codesandbox.io/s/sleepy-dan-ghlh1u?file=/pages/_app.js

**Comment by gaearon** [comment]
Minimal repro without Next.js: https://codesandbox.io/s/compassionate-night-uzyxv1?file=/src/App.js

**Comment by gaearon** [comment]
Seems like a bug. Weird that I can't repro this in tests.

**Comment by xiel** [comment]
Please see my comment in the PR: https://github.com/facebook/react/pull/24404#issuecomment-1104798773

**Comment by gurkerl83** [comment]
@gnoff @gaearon 

In my opinion, when a suspender alike component is involved, it is similar if you take direct action patching the DOM on the client-side, within your own code, because React does not do this anymore.

```
<Suspense>
<CompA><div>A</div>
<CompB - [Suspender]><div>B</div>
<CompC><div>C</div>
</Suspense>
```

- **1st render on the server**
```
<div>A</div>
<div>B - when the suspenders promise can get resolved on the server otherwise initiate a fallback and resolve at client-side, the final tree is unknown</div>
<div>C if the promise within the suspender does resolve on the server; unknown otherwise, the output depends on the result of the promise</div> 
```

- **1st render on the client**
If the suspense boundary does not resolve on the server completely producing the final tree, client-side render kicks in; the client re-runs the entire tree within the suspense block again and I think this is happening here.

- **BUT**
**The client will get aborted immediately because the erroring mechanism gets enforced before the actual thing can produce something**. See the **comments** made in the erroring sections in 
- tryToClaimNextHydratableInstance (the erroring in here seems to be what is producing this whole cascade of problems, **no erroring was enforced before, this was brought in a couple of month back**) and 
- popHydrationState (erroring in here seems now reasonable, before Reacts DOM-Patching was handled here) 
- functions extracted from Reacts source code.

Maybe this helps...

**Comment by gnoff** [comment]
@xiel can you confirm the fix in #24480 addresses your concern. Here is a slightly modified version of the original repro on codesandbox using the build from that PR and I believe it is doing what you expect: https://codesandbox.io/s/trusting-buck-2j21to?file=/package.json

**Comment by xiel** [comment]
@gnoff Looks good, thanks so much for working on this! 🤟

**Comment by antonybudianto** [comment]
Hi, will the fix be on 18.1.1 ? I see the issue still persist on 18.1.0
however I tried `https://react-builds.vercel.app/api/prs/24480/packages/react` and it solved on my repo.

**Comment by gaearon** [comment]
The plan is to release it in 18.2.

**Comment by hdlion** [comment]
Hi, Issue is fixed in react 18.2.0.

`npm run react@18.2.0-next-47944142f-20220608
npm run react-dom@18.2.0-next-47944142f-20220608`

but lazy component is mounting twice. Console.log in componentDidMount is calling twice, As we will migrate class component to functional component later on.

**Comment by gaearon** [comment]
It's expected that there is a development-only remount if your code is running in Strict Mode.

https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-strict-mode
